cpp/src/utilities/prefetch.cpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2020-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/prefetch.hpp>
+
+#include <rmm/cuda_device.hpp>
+
+#include <iostream>
+
+namespace cudf::experimental::prefetch {
+
+namespace detail {
+
+PrefetchConfig& PrefetchConfig::instance()
+{
+  static PrefetchConfig instance;
+  return instance;
+}
+
+bool PrefetchConfig::get(std::string_view key)
+{
+  // Default to not prefetching
+  if (config_values.find(key.data()) == config_values.end()) {
+    return (config_values[key.data()] = false);
+  }
+  return config_values[key.data()];
+}
+void PrefetchConfig::set(std::string_view key, bool value) { config_values[key.data()] = value; }
+
+cudaError_t prefetch_noexcept(std::string_view key,
+                              void const* ptr,
+                              std::size_t size,
+                              rmm::cuda_stream_view stream,
+                              rmm::cuda_device_id device_id) noexcept
+{
+  if (PrefetchConfig::instance().get(key)) {
+    if (PrefetchConfig::instance().debug) {
+      std::cerr << "Prefetching " << size << " bytes for key " << key << " at location " << ptr
+                << std::endl;
+    }
+    return cudaMemPrefetchAsync(ptr, size, device_id.value(), stream.value());
+  }
+  return cudaSuccess;
+}
+
+void prefetch(std::string_view key,
+              void const* ptr,
+              std::size_t size,
+              rmm::cuda_stream_view stream,
+              rmm::cuda_device_id device_id)
+{
+  auto result = prefetch_noexcept(key, ptr, size, stream, device_id);
+  // Ignore cudaErrorInvalidValue because that will be raised if
+  // prefetching is attempted on unmanaged memory.
+  if ((result != cudaErrorInvalidValue) && (result != cudaSuccess)) {
+    std::cerr << "Prefetch failed" << std::endl;
+    CUDF_CUDA_TRY(result);
+  }
```

Comment:
This still makes me uncomfortable only because in #16265 I had to add a `cudaGetLastError()` call to clear the invalid-value error otherwise it would show up in a later CUDA call. I'm not sure why that is not occurring with this PR. I suspect we are careful to enable prefetching only on managed memory. But this check implies we don't want it to fail on unmanaged memory as well.

```suggestion
  }
  if (result == cudaErrorInvalidValue) { result = cudaGetLastError(); }
```

I did a quick test on my local machine with unmanaged memory and the `cudaErrorInvalidValue` causes a failure later in the code without the suggested line.

Reply:
Oh right, I forgot to add this. Good idea. Presumably we want to put this in the noexcept version instead of here, right?

Reply:
I've pushed a slightly different version of this change that should accomplish what we want. Let me know if you agree.